### PR TITLE
Add asyncio support for LLM (OpenAI), Chain (LLMChain, LLMMathChain), and Agent

### PR DESCRIPTION
**Pull request by agola11**
Supporting asyncio in langchain primitives allows for users to run them concurrently and creates more seamless integration with asyncio-supported frameworks (FastAPI, etc.)

Summary of changes:

**LLM**
* Add `agenerate` and `_agenerate`
* Implement in OpenAI by leveraging `client.Completions.acreate`

**Chain**
* Add `arun`, `acall`, `_acall`
* Implement them in `LLMChain` and `LLMMathChain` for now

**Agent**
* Refactor and leverage async chain and llm methods
* Add ability for `Tools` to contain async coroutine
* Implement async SerpaPI `arun`

Create demo notebook.

Open questions:
* Should all the async stuff go in separate classes? I've seen both patterns (keeping the same class and having async and sync methods vs. having class separation)

### CODE REVIEW

File: langchain/agents/agent.py [hunk 1]
```diff
@@ -399,3 +446,77 @@ def _call(self, inputs: Dict[str, str]) -> Dict[str, Any]:
...
+                    {"name": str(tool.func)[:60] + "..."},
+                    output,
+                    verbose=self.verbose,
+                )
+                try:
+                    # We then call the tool on the tool input to get an observation
+                    observation = (
+                        await tool.coroutine(output.tool_input)
+                        if tool.coroutine
+                        else tool.func(output.tool_input)
```
Comment (nfcampos) [inline langchain/agents/agent.py:491]: This should probably be called with one of the asyncio methods for executing sync functions in another thread. If left like this it blocks the event loop for the duration of the tool call

File: langchain/agents/agent.py [hunk 2]
```diff
@@ -399,3 +446,77 @@ def _call(self, inputs: Dict[str, str]) -> Dict[str, Any]:
...
+                    )
+                    color = color_mapping[output.tool]
+                    return_direct = tool.return_direct
+                except (KeyboardInterrupt, Exception) as e:
+                    self.callback_manager.on_tool_error(e, verbose=self.verbose)
```
Comment (nfcampos) [inline langchain/agents/agent.py:496]: Should the callback manager have async methods? If I'm using an async agent I might also want to use an async callback handler

File: langchain/chains/llm.py [hunk 3]
```diff
@@ -71,12 +85,20 @@ def generate(self, input_list: List[Dict[str, Any]]) -> LLMResult:
...
-        response = self.llm.generate(prompts, stop=stop)
-        return response
+        return prompts, stop
 
     def apply(self, input_list: List[Dict[str, Any]]) -> List[Dict[str, str]]:
         """Utilize the LLM generate method for speed gains."""
         response = self.generate(input_list)
+        return self.create_outputs(response)
+
+    async def aapply(self, input_list: List[Dict[str, Any]]) -> List[Dict[str, str]]:
```
Comment (nfcampos) [inline langchain/chains/llm.py:95]: An unfortunate name ha, but I guess it is what it is

File: langchain/serpapi.py [hunk 4]
```diff
@@ -75,45 +107,43 @@ def validate_environment(cls, values: Dict) -> Dict:
...
+        def construct_url_and_params() -> Tuple[str, Dict[str, str]]:
+            params = self.get_params(query)
+            params["source"] = "python"
+            if self.serpapi_api_key:
+                params["serp_api_key"] = self.serpapi_api_key
+            params["output"] = "json"
+            url = "https://serpapi.com/search"
+            return url, params
+
+        async with aiohttp.ClientSession() as session:
```
Comment (nfcampos) [inline langchain/serpapi.py:122]: This is not recommended, sessions should be reused if possible

**Comment by nfcampos** [comment]
Making the async methods live in the same class makes reusing helper methods across both implementations easier

File: langchain/agents/agent.py (hunk 1 above)
Comment (nfcampos) [reply langchain/agents/agent.py:491]:
```py
asyncio.get_event_loop().run_in_executor(None, tool.func, output.tool_input)
```

File: langchain/serpapi.py (hunk 4 above)
Comment (nfcampos) [reply langchain/serpapi.py:122]: Also, the docs for this feature should probably mention this part of the openai python docs https://github.com/openai/openai-python#async-api

File: langchain/agents/agent.py (hunk 1 above)
Comment (agola11) [reply langchain/agents/agent.py:491]: Yep, great point

File: langchain/agents/agent.py (hunk 2 above)
Comment (agola11) [reply langchain/agents/agent.py:496]: Rn, callbacks are all synchronous but on our roadmap to change soon

File: langchain/serpapi.py (hunk 4 above)
Comment (agola11) [reply langchain/serpapi.py:122]: will do -- thanks

**Review by nfcampos — COMMENTED** [review]
Looks good

**Review by hwchase17 — APPROVED** [review]

**Comment by tmbo** [comment]
How is this compatible with the global `OpenAICallbackHandler`? e.g. let's say you have multiple of the contexts mentioned in https://langchain.readthedocs.io/en/latest/modules/llms/examples/token_usage_tracking.html 

```python
import time
import asyncio
from langchain.callbacks import get_openai_callback
from langchain.llms import OpenAI

async def async_generate(llm):
    with get_openai_callback() as cb:
        resp = await llm.agenerate(["Hello, how are you?"])
        print(resp.generations[0][0].text)
        # not consistent as async execution will share the global handler
        print(cb.total_tokens)

async def generate_concurrently():
    llm = OpenAI(temperature=0.9)
    tasks = [async_generate(llm) for _ in range(10)]
    await asyncio.gather(*tasks)
```

I am not sure if this implementation is complete without handling callbacks properly

**Pull request merged**